File: game_cpp/src/barrier.cpp
#include "barrier.h"


barrier::barrier(float* coeff, int _degree, float _xmin, float _xmax, float prec, int padd, POINT origin)
{
	coefficients = coeff;
	degree = _degree;
	padding = padd;
	precision = prec;
	xmin = _xmin;
	xmax = _xmax;
	view_loc = origin;
	part_loc = winPoint_to_myPoint(origin);
	alive = true;
}
barrier::~barrier()
{
	delete coefficients;
}



bool barrier::isCollision(POINT test_pt)
{
	//Prelimnary: convert test_pt to non-translated coordinates.
	POINT test_pt_prime;
	test_pt_prime.x = test_pt.x - view_loc.x;
	test_pt_prime.y = test_pt.y - view_loc.y;



	//First, find the point, curve_pt, which defines the shortest secant line (perp. to tan.) of the curve that passes through the test point.
	POINT curve_pt;
	bool found = false;

	float error_margin = 200;
	float prec = .1;
	for (float x = xmin; (x < xmax); x += prec)
	{
		//Derivative of the polynomial
		float dydx = getDerivative(x);
		float y = f_of_x(x);

		float diff = dydx + (x - test_pt_prime.x) / (y - test_pt_prime.y);
		if (abs(diff) < .1)
		{
			//Close enough to the perpendicular line we seek.

			curve_pt.x = x;
			curve_pt.y = y;
			found = true;
			break;
		}
	}


	//Next, check length of line from test_pt to curve_pt.
	if (found)
	{
		float dist = sqrt(pow((curve_pt.x - test_pt_prime.x), 2) + pow((curve_pt.y - test_pt_prime.y), 2));

		if (dist <= padding)
		{
			return true;
		}
	}
	return false;

}

float barrier::getDerivative(float x)
{
	float* curr = coefficients;
	float dydx = 0;
	for (int i = 0; i <= degree; i++)
	{
		dydx += *curr * i * pow(x, i - 1);
		curr++;
	}
	return dydx;
}

float barrier::f_of_x(float x)
{
	//Maybe in future, update to return function value in pointer variable passed as argument and set return value of this function to a bool indicating whether or not the function is defined at the parameter given.

	float* curr = coefficients;
	float y = 0;
	int i = 0;
	while (i <= degree)
	{
		y += *curr * pow(x, i);
		i++;
		curr++;
	}
	return y;
}


void barrier::draw(HDC hdc, HWND hwnd)
{
	std::vector<POINT> pts;

	//note: using a scaling factor might be advisable in future.
	
	
	for (float x = xmin; x < xmax; x += precision)
	{
		float* curr = coefficients;
		float y = 0;
		int i = 0;
		while (i<= degree)
		{
			y += *curr * pow(x, i);
			i++;
			curr++;
		}
		POINT pt;
		pt.x = x + view_loc.x;
		pt.y = y + view_loc.y;
		pts.push_back(pt);
	}
	HPEN pen = CreatePen(PS_SOLID, 3, COLORREF(0x000000));
	SelectObject(hdc, pen);
	MoveToEx(hdc, pts[0].x, pts[0].y, NULL);
	for (int i = 1; i < pts.size(); i++)
	{
		LineTo(hdc, pts[i].x, pts[i].y);
	}
	DeleteObject(pen);
}

File: sim/barrier.ts
import { Entity } from './entity'
import { EntityState } from './types'

export class Barrier extends Entity {
  coefficients: number[]
  degree: number
  xMin: number; xMax: number
  precision: number
  padding: number
  originX: number; originY: number

  constructor(
    coefficients: number[],
    degree: number,
    xMin: number, xMax: number,
    precision: number,
    padding: number,
    originX: number, originY: number,
  ) {
    super('barrier')
    this.coefficients = coefficients
    this.degree = degree
    this.xMin = xMin; this.xMax = xMax
    this.precision = precision
    this.padding = padding
    this.originX = originX; this.originY = originY
    this.x = originX; this.y = originY
    this.mass = 500
    this.ignoreCollision = true
    this.ignoreElectricity = true
    this.gravityException = true
  }

  // Evaluate polynomial at x: sum of coefficients[i] * x^i
  f(x: number): number {
    let y = 0
    for (let i = 0; i <= this.degree; i++) {
      y += this.coefficients[i] * Math.pow(x, i)
    }
    return y
  }

  getDerivative(x: number): number {
    let dydx = 0
    for (let i = 1; i <= this.degree; i++) {
      dydx += this.coefficients[i] * i * Math.pow(x, i - 1)
    }
    return dydx
  }

  // Translates test point to barrier-local coordinates, then searches for
  // the curve point closest (perpendicular) to the test point.
  // Matches C++ barrier::isCollision() logic exactly.
  isCollision(px: number, py: number): boolean {
    const lx = px - this.originX
    const ly = py - this.originY

    const prec = 0.1
    let found = false
    let curvePx = 0, curvePy = 0

    for (let x = this.xMin; x < this.xMax; x += prec) {
      const dydx = this.getDerivative(x)
      const y = this.f(x)
      const diff = dydx + (x - lx) / (y - ly)
      if (Math.abs(diff) < 0.1) {
        curvePx = x
        curvePy = y
        found = true
        break
      }
    }

    if (found) {
      const dx = curvePx - lx
      const dy = curvePy - ly
      const dist = Math.sqrt(dx * dx + dy * dy)
      return dist <= this.padding
    }
    return false
  }

  // Returns evenly-spaced sample points along the curve (for frontend rendering)
  getCurvePoints(step = 1): Array<{ x: number; y: number }> {
    const pts: Array<{ x: number; y: number }> = []
    for (let x = this.xMin; x < this.xMax; x += step) {
      pts.push({ x: x + this.originX, y: this.f(x) + this.originY })
    }
    return pts
  }

  override toJSON(): EntityState {
    return { ...super.toJSON(), type: 'barrier' }
  }
}
